### ISSUE TITLE
Panicking in a callback with feature parallel enabled causes quickjs to trigger assertion

### ISSUE THREAD

**Issue by DelSkayn** [DelSkayn/rquickjs#11]
Quickjs has an assertion which checks if all javascript objects are freed before destroying the runtime.
If a new callback function created from rust panics this assertion is triggered if rquickjs is compiled with the parallel feature enabled.
This probably means that JSValue is not properly dropped somewhere.
However code dealing with drop is mostly the same with the parallel feature enabled and without, which makes the fact that this bug only happens in parallel rather confusing.

Minimal example:
```rust
use rquickjs::{Context, Function, Runtime};

fn main() {
    let rt = Runtime::new().unwrap();
    let ctx = Context::full(&rt).unwrap();
    ctx.with(|ctx| {
        println!("before function");
        let f = Function::new::<_, _, _, (), _>(ctx, "test", move |_, _: (), _: ()| {
            panic!()
            //Ok(())
        })
        .unwrap();
        f.call::<_, ()>(()).unwrap();
    })
//assertion triggered here:
//8043/out/quickjs.c:1979: JS_FreeRuntime: Assertion `list_empty(&rt->gc_obj_list)' failed.
}
```

**Comment by DelSkayn** [comment]
Bug was not pressent in 7346670be8d9c16bee879b9d82f5cc50e1b56e39

**Comment by DelSkayn** [comment]
Bug seems to be introduced with new quickjs version

**Comment by katyo** [comment]
Have you any ideas how to solve it?

**Comment by DelSkayn** [comment]
I have been debugging it for a while but I am not sure what is happening.

The bug can be reproduced by only switching out `Ref`'s `RefCell` for a `Mutex`. 
I have no idea why this difference would results in different behaviour from quickjs.

Because of the error being seemingly separated from the assertion I think it might happen because a value on the stack might be accessed wrong and the difference in guard data on the stack between the `RefCell` and `Mutex` might be what causes the difference.

But I have not been able to find anything.

**Comment by DelSkayn** [comment]
Wow this one throw me for a loop, fixed in 32e98225fbee7242acf09d34977d0c7383d52b80.
